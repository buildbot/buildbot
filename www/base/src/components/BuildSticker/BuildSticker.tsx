/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import './BuildSticker.scss';
import {observer} from 'mobx-react';
import {Link} from 'react-router-dom';
import {Card} from 'react-bootstrap';
import {Build, Builder, results2class, results2text} from 'buildbot-data-js';
import {BadgeStatus, durationFormat, useCurrentTime} from 'buildbot-ui';

type BuildStickerProps = {
  build: Build;
  builder: Builder;
};

export const BuildSticker = observer(({build, builder}: BuildStickerProps) => {
  const now = useCurrentTime();

  return (
    <Card className={'bb-buildsticker ' + results2class(build, null)}>
      <Card.Body>
        <div className="bb-buildsticker-left">
          <BadgeStatus className={results2class(build, null)}>{results2text(build)}</BadgeStatus>
          <Link to={`builders/${builder.builderid}/builds/${build.number}`}>
            {builder.name}/{build.number}
          </Link>
        </div>
        <div className="bb-buildsticker-left">
          <span className="bb-buildsticker-time">
            {durationFormat((build.complete ? build.complete_at! : now) - build.started_at)}
          </span>
          <span>{build.state_string}</span>
        </div>
      </Card.Body>
    </Card>
  );
});
