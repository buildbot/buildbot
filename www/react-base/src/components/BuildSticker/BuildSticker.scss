@import "../../../../react-ui/dist/colors.scss";

.bb-buildsticker {
  width: 100%;
  margin: 0 2px 2px 0;

  .bb-buildsticker-left {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    margin-left: 0;
    margin-right: 0;

    > .bb-badge-status, > .bb-buildsticker-time {
      float: right;
    }
  }

  &.results_SUCCESS {
    border-color: $bb-color-border-results-success;
  }

  &.results_WARNINGS {
    border-color: $bb-color-border-results-warnings;
  }

  &.results_FAILURE {
    border-color: $bb-color-border-results-failure;
  }

  &.results_SKIPPED {
    border-color: $bb-color-border-results-skipped;
  }

  &.results_EXCEPTION {
    border-color: $bb-color-border-results-exception;
  }

  &.results_RETRY {
    border-color: $bb-color-border-results-retry;
  }

  &.results_CANCELLED {
    border-color: $bb-color-border-results-cancelled;
  }

  &.results_PENDING {
    border-color: $bb-color-border-results-pending;
  }

  &.results_UNKNOWN {
    border-color: $bb-color-border-results-unknown;
  }
}
