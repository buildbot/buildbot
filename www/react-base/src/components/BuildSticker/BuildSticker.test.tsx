/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import renderer from 'react-test-renderer';
import {Build, Builder, FAILURE, SUCCESS} from "buildbot-data-js";
import {BuildSticker} from "./BuildSticker";
import {MemoryRouter} from "react-router-dom";
import {TimeContext, TimeStore} from "buildbot-ui";

function assertBuildStickerRenderSnapshot(build: Build, builder: Builder) {
  const timeStore = new TimeStore();
  timeStore.setTimeFromString("2022-01-01T00:00:00.000Z");

  const component = renderer.create(
    <MemoryRouter>
      <TimeContext.Provider value={timeStore}>
        <BuildSticker build={build} builder={builder}/>
      </TimeContext.Provider>
    </MemoryRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
}

describe('buildsticker component', function() {
  it('simple', () => {
    const build: Build = {
      buildid: 3,
      builderid: 2,
      number: 1,
      complete: false,
      started_at: null,
      results: -1,
      state_string: null,
    } as any;

    const builder: Builder = {
      builderid: 2
    } as any;

    assertBuildStickerRenderSnapshot(build, builder);
  });

  it('pending', () => {
    const build: Build = {
      buildid: 3,
      builderid: 2,
      number: 1,
      complete: false,
      started_at: 20,
      results: -1,
      state_string: 'pending',
    } as any;

    const builder: Builder = {
      builderid: 2
    } as any;

    assertBuildStickerRenderSnapshot(build, builder);
  });

  it('success', () => {
    const build: Build = {
      buildid: 3,
      builderid: 2,
      number: 1,
      complete: true,
      started_at: 20,
      complete_at: 30,
      results: SUCCESS,
      state_string: 'finished',
    } as any;

    const builder: Builder = {
      builderid: 2
    } as any;

    assertBuildStickerRenderSnapshot(build, builder);
  });

  it('failed', () => {
    const build: Build = {
      buildid: 3,
      builderid: 2,
      number: 1,
      complete: true,
      started_at: 20,
      complete_at: 30,
      results: FAILURE,
      state_string: 'failed',
    } as any;

    const builder: Builder = {
      builderid: 2
    } as any;

    assertBuildStickerRenderSnapshot(build, builder);
  });
});
