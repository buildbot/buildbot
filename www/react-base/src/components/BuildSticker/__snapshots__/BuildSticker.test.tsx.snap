// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildsticker component failed 1`] = `
<div
  className="bb-buildsticker results_FAILURE card"
>
  <div
    className="card-body"
  >
    <div
      className="bb-buildsticker-left"
    >
      <div
        className="bb-badge-status results_FAILURE"
      >
        FAILURE
      </div>
      <a
        href="/builders/2/builds/1"
        onClick={[Function]}
      >
        /
        1
      </a>
    </div>
    <div
      className="bb-buildsticker-left"
    >
      <span
        className="bb-buildsticker-time"
      >
        10 s
      </span>
      <span>
        failed
      </span>
    </div>
  </div>
</div>
`;

exports[`buildsticker component pending 1`] = `
<div
  className="bb-buildsticker results_PENDING card"
>
  <div
    className="card-body"
  >
    <div
      className="bb-buildsticker-left"
    >
      <div
        className="bb-badge-status results_PENDING"
      >
        ...
      </div>
      <a
        href="/builders/2/builds/1"
        onClick={[Function]}
      >
        /
        1
      </a>
    </div>
    <div
      className="bb-buildsticker-left"
    >
      <span
        className="bb-buildsticker-time"
      >
        18992 days 23:59:40
      </span>
      <span>
        pending
      </span>
    </div>
  </div>
</div>
`;

exports[`buildsticker component simple 1`] = `
<div
  className="bb-buildsticker results_UNKNOWN card"
>
  <div
    className="card-body"
  >
    <div
      className="bb-buildsticker-left"
    >
      <div
        className="bb-badge-status results_UNKNOWN"
      >
        ...
      </div>
      <a
        href="/builders/2/builds/1"
        onClick={[Function]}
      >
        /
        1
      </a>
    </div>
    <div
      className="bb-buildsticker-left"
    >
      <span
        className="bb-buildsticker-time"
      >
        18993 days 0:00:00
      </span>
      <span />
    </div>
  </div>
</div>
`;

exports[`buildsticker component success 1`] = `
<div
  className="bb-buildsticker results_SUCCESS card"
>
  <div
    className="card-body"
  >
    <div
      className="bb-buildsticker-left"
    >
      <div
        className="bb-badge-status results_SUCCESS"
      >
        SUCCESS
      </div>
      <a
        href="/builders/2/builds/1"
        onClick={[Function]}
      >
        /
        1
      </a>
    </div>
    <div
      className="bb-buildsticker-left"
    >
      <span
        className="bb-buildsticker-time"
      >
        10 s
      </span>
      <span>
        finished
      </span>
    </div>
  </div>
</div>
`;
